fix: Allow user to set custom endpoints

danieljbruce opened the pull request:
This change allows the user to set a custom endpoint when creating the client and it will work with the backend the way it is supposed to. It also ensures that we can still work with the emulator using the client.

danieljbruce committed:
Write a test for custom endpoints

With the code in its present state, this test should fail. A fix for regional endpoints intends to clean this up.

danieljbruce committed:
Limit setting credentials to insecure

Only set the credentials to insecure if the custom endpoint is localhost and credentials are not provided. This way the user can set custom endpoints with the client and the request will work.

danieljbruce committed:
Change the tests for custom endpoints

There is a test that uses a mock which is preventing the base url from being set. The test should be more realistic and when it is, the test passes with the new change.

alexander-fenster commented on src/index.ts:
```diff
@@ -504,8 +504,12 @@ class Datastore extends DatastoreRequest {
+    const isUsingEmulator =
+      this.baseUrl_ &&
+      (this.baseUrl_.includes('localhost') ||
+        this.baseUrl_.includes('127.0.0.1'));
```
Just in case, maybe add `::1` as an IPv6 version of localhost?

danieljbruce replied:
Done

danieljbruce committed:
Add another localhost option

There is a third way the user can specify localhost. We should include this way as an option.

edance commented:
hey all! This PR broke the docker setup for us. We're running the datastore emulator using a docker container so the datastore emulator host is something like `datastore:8081`. We want a connection that is both insecure and not to localhost.

How can we make changes to this code? If `process.env.DATASTORE_EMULATOR_HOST` is set, should we make `isUsingEmulator == true`
